### PR TITLE
Bug 770785: delete client record when Android Account is deleted.

### PR DESCRIPTION
**Pull request by ncalexan**

### CODE REVIEW

**Comment by ncalexan** [comment]
I built this on top of the auth changes since it needs to get Sync account parameters from an Account object.

**Comment by ncalexan** [comment]
Test instructions at https://bugzilla.mozilla.org/show_bug.cgi?id=770785#c3.

**Comment by ncalexan** [comment]
Fixed license and rebased to develop.

File: src/main/java/org/mozilla/gecko/sync/config/ClientRecordTerminator.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+import org.mozilla.gecko.sync.GlobalSession;
+import org.mozilla.gecko.sync.Logger;
+import org.mozilla.gecko.sync.net.BaseResource;
+import org.mozilla.gecko.sync.net.SyncStorageRecordRequest;
+import org.mozilla.gecko.sync.net.SyncStorageRequestDelegate;
+import org.mozilla.gecko.sync.net.SyncStorageResponse;
+
+/**
+ * Bug XXX: when an Android Account is deleted, we need to (try to) delete the
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/config/ClientRecordTerminator.java]: Bug number, plz.

File: src/main/java/org/mozilla/gecko/sync/config/ClientRecordTerminator.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+      final String clusterURL,
+      final String clientGuid)
+    throws Exception {
+
+    // Would prefer to delegate to SyncConfiguration, but that would proliferate static methods.
+    final String collection = "clients";
+    final URI wboURI = new URI(clusterURL + GlobalSession.API_VERSION + "/" + username + "/storage/" + collection + "/" + clientGuid);
+
+    final SyncStorageRecordRequest r = new SyncStorageRecordRequest(wboURI);
+    r.delegate = new SyncStorageRequestDelegate() {
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/config/ClientRecordTerminator.java]: We really need a self-contained client library, don't we?

File: src/test/java/org/mozilla/android/sync/test/integration/TestClientRecordTerminator.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  static final String TEST_USERNAME     = "6gnkjphdltbntwnrgvu46ey6mu7ncjdl";
+  static final String TEST_PASSWORD     = "test0425";
+  static final String TEST_USER_PASS    = TEST_USERNAME + ":" + TEST_PASSWORD;
+  static final String TEST_SYNC_KEY     = "fuyx96ea8rkfazvjdfuqumupye"; // Weave.Identity.syncKey
+
+  static final String TEST_CLIENT_GUID_PREFIX  = "testGuid";
+
+  private KeyBundle syncKeyBundle;
+  private MockGlobalSessionCallback callback;
+  // We go to the trouble of using a session and it's internal SyncConfiguration
```
Comment (rnewman) [inline src/test/java/org/mozilla/android/sync/test/integration/TestClientRecordTerminator.java]: s/it's/its

File: src/main/java/org/mozilla/gecko/sync/config/ClientRecordTerminator.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+          Logger.warn(LOG_TAG, "Server error message was: " + response.getErrorMessage());
+        } catch (Exception e) {
+          // Do nothing.
+        }
+        BaseResource.consumeEntity(response);
+      }
+
+      @Override
+      public void handleRequestError(Exception ex) {
+        Logger.error(LOG_TAG, "Got exception trying to delete client record with GUID " + clientGuid + " from server.", ex);
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/config/ClientRecordTerminator.java]: `getAccountRemovalAllowed` is permitted to throw on a network error "if the authenticator could not honor the request due to a network error". Think about whether we care enough. I'm erring on the side of "we don't care if it fails", but you might disagree.

**Comment by rnewman** [comment]
r+.

**Comment by ncalexan** [comment]
Rebased to develop; addressed review comments; will merge after device testing.

**Pull request merged**